### DISCUSSION TITLE
codeql build command

### DISCUSSION THREAD

**Discussion by sam-cha-tfs** [github/codeql#12962] [General]
I have a build script that I pass into the --command option. The script needs to take in parameters. How do I specify these parameters?

`codeql database create project-db --language=csharp --command=c:\\build.py <parameter1>`

**Comment by hmakholm** [comment]
If the arguments are simple and don't need quoting of their own, you can quote the entire `"--command=name arg arg arg"` flag.

CodeQL will then do its own argument splitting and possibly re-quote the argument when it tries to start the command. This argument splitting is not particularly well-specified, so if you have arguments that need quoting or escaping of their own, our usual recommendation is to hard-code them in the script -- possibly write a wrapper script on the fly that calls the actual build script with the desired arguments.

**Reply by sam-cha-tfs** [reply]
Below is what I have so far but it seems like codeql is not parsing the arguments correctly.

`codeql database create project-db --language=csharp --command="c:\\build.pl --file="<filename>" --config="<config name>"  --hash="" "`

**Reply by hmakholm** [reply]
Right, if simple splitting on spaces is not enough, then your best bet is to create a wrapper script of your own.

**Reply by sam-cha-tfs** [reply]
Is this build output indicating that there are two build commands?

`Running build command: [perl, build.pl]`

**Reply by hmakholm** [reply]
It indicates there is one build command which has been split into two words.

Internally, `codeql` assumes the Unix model of command lines where a command consists of an array of strings, which are passed to the new process one by one. This doesn't quite match the native model of Windows (where a command line is a single string and it's up to the new process to split it into words if it wants to), which is one of the underlying causes why giving a single `--command` option doesn't always work well.

**Reply by sam-cha-tfs** [reply]
It seems like codeql does not like to parse empty strings such as ""  as part of the build command. Even using escape string doesn't work. I assume the best option would be to use a script.
